<h1 align="center">Física Computacional.</h1>
<h1 align="center">Semestre 2015-2</h1>

<h2>Sergio A. Alcalá Corona </h2> 

---
### Rodrigo Vega Vilchis
### Daniel Rojo Mata
---

<h1 align="center">Programación para la física computacional</h1> 

## Práctica 1.  Programación básica


### Programación básica para la física

---

### **EJERCICIOS**:

$\;$

#### **1. Altitud de un satélite**: se va a lanzar un satélite en una órbita circular alrededor de la Tierra de modo que orbite el planeta una vez cada $T$ segundos.

#### **(a)** Demuestre que la altitud $h$ sobre la superficie de la Tierra que debe tener el satélite es:

<h3 align="center">$h = \biggl( {GMT^2\over4\pi^2} \biggr)^{1/3} - R$,</h3> 

donde $G = 6,67 × 10^{−11} m^3 kg^{−1}s^{−2}$ es la constante gravitatoria de Newton, $M = 5,97 ×
1024 kg$ es la masa de la Tierra, y $R = 6371 km$ es su radio.




#### **(b)** Escribe un programa que le pida al usuario que ingrese el valor deseado de $T$ y luego calcule e imprima la altura correcta en metros.


In [1]:
import math

# Constantes
G = 6.67e-11   # m^3 kg^-1 s^-2
M = 5.97e24    # kg
R = 6371e3     # m

# Pedir al usuario el tiempo T (en segundos)
T = float(input("Ingresa el periodo orbital T en segundos: "))

# Cálculo de h
h = (G * M * T**2 / (4 * math.pi**2))**(1/3) - R

# Mostrar resultado
print("La altitud del satélite es:", h, "metros")



Ingresa el periodo orbital T en segundos: 23
La altitud del satélite es: -6196257.437492619 metros


#### **(c)** Utiliza tu programa para calcular las altitudes de los satélites que orbitan la Tierra una vez al día (la llamada *órbita "geosincrónica"*), una vez cada **90 minutos** y una vez cada **45 minutos**. 
#### ¿Qué concluyes del último de estos cálculos?

In [2]:
# Parte c) Usar el programa para tres valores específicos
print("\nCalculando altitudes para distintos periodos...")

T_dia = 24 * 3600
T_90min = 90 * 60
T_45min = 45 * 60

lista_T = [T_dia, T_90min, T_45min]
nombres = ["Geoestacionaria (24h)", "Cada 90 minutos", "Cada 45 minutos"]

for i in range(3):
    T = lista_T[i]
    h = (G * M * T**2 / (4 * math.pi**2))**(1/3) - R
    print(nombres[i], "-> Altitud:", h, "metros")


Calculando altitudes para distintos periodos...
Geoestacionaria (24h) -> Altitud: 35855910.176174976 metros
Cada 90 minutos -> Altitud: 279321.6253728606 metros
Cada 45 minutos -> Altitud: -2181559.8978108233 metros


#### **2. Relatividad especial:** Una nave espacial viaja desde la Tierra en línea recta a una velocidad relativista $v$ a otro planeta a $x$ años luz de distancia. 

#### Escribe un programa que le pida al usuario el valor de $x$ y la velocidad $v$ como una fracción de la velocidad de la luz $c$, y que imprima el tiempo en años que tarda la nave espacial en llegar a su destino

#### **(a)** en el marco de reposo de un observador en la Tierra y

In [3]:
import math

# c: velocidad de la luz en unidades de años luz por año
c = 1.0

# Solicitar entrada del usuario
x = float(input("Ingresa la distancia al planeta (en años luz): "))
v_frac = float(input("Ingresa la velocidad como fracción de c (por ejemplo, 0.99): "))

# Calcular tiempos
v = v_frac * c

if v >= c:
    print("La velocidad debe ser menor que la velocidad de la luz.")
else:
    t_tierra = x / v
    t_nave = t_tierra * math.sqrt(1 - v**2 / c**2)

    print("\nTiempo según un observador en la Tierra:", t_tierra, "años")



Ingresa la distancia al planeta (en años luz): 23
Ingresa la velocidad como fracción de c (por ejemplo, 0.99): 0.8

Tiempo según un observador en la Tierra: 28.75 años


#### **(b)** como lo percibiría un pasajero a bordo de la nave. 

In [4]:
import math

# c: velocidad de la luz en unidades de años luz por año
c = 1.0

# Solicitar entrada del usuario
x = float(input("Ingresa la distancia al planeta (en años luz): "))
v_frac = float(input("Ingresa la velocidad como fracción de c (por ejemplo, 0.99): "))

# Calcular tiempos
v = v_frac * c

if v >= c:
    print("La velocidad debe ser menor que la velocidad de la luz.")
else:
    t_tierra = x / v
    t_nave = t_tierra * math.sqrt(1 - v**2 / c**2)

    print("Tiempo según un pasajero en la nave:", t_nave, "años")


Ingresa la distancia al planeta (en años luz): 23
Ingresa la velocidad como fracción de c (por ejemplo, 0.99): 0.8
Tiempo según un pasajero en la nave: 17.249999999999996 años


#### Usa tu programa para calcular las respuestas para un planeta a 10 años luz de distancia con $v = 0.99c$

In [5]:

# Parte extra: ejemplo para 10 años luz y v = 0.99c
print("\nEjemplo con x = 10 años luz, v = 0.99c")
x = 10
v = 0.99
t_tierra = x / v
t_nave = t_tierra * math.sqrt(1 - v**2)
print("Tiempo en Tierra:", t_tierra, "años")
print("Tiempo en la nave:", t_nave, "años")



Ejemplo con x = 10 años luz, v = 0.99c
Tiempo en Tierra: 10.1010101010101 años
Tiempo en la nave: 1.424922826228878 años


#### **3. Órbitas planetarias**: En el espacio, la órbita de un cuerpo alrededor de otro (como un planeta alrededor del Sol), no necesariamente es circular. En general, toma la forma de una elipse, con el cuerpo a veces más cerca y otras más lejos. Si tenemos la distancia $\ell_1$ de máxima aproximación de un planeta al Sol (su \textit{perihelio}), y su velocidad lineal $v_1$ en el perihelio, entonces cualquier otra propiedad de la órbita se puede calcular a partir de estas dos cantidades de la siguiente manera:

#### **(a)** La segunda ley de Kepler nos dice que la distancia $\ell_2$ y la velocidad $v_2$ del planeta en su punto más distante, o \textit{afelio}, satisfacen que $\ell_2 v_2 = \ell_1 v_1$.  Al mismo tiempo, la energía total, cinética más la gravitatoria, de un planeta con velocidad $v$ y distancia $r$ del Sol está dada por:

<h3 align="center">$E = \frac{1}{2} m v^2 - G {mM\over r}$,</h3>
       
#### donde $m$ es la masa del planeta, $M=1.9891\times10^{30}\,$kg es la masa del Sol y $G=6.6738\times10^{-11}\,\mathrm{m^3\,kg^{-1}\,s^{-2}}$ es la constante gravitatoria de Newton. Dado que la energía debe conservarse, demuestre que $v_2$ es la raíz más pequeña de la ecuación cuadrática:

<h3 align="center">$v_2^2 - {2GM\over v_1\ell_1} v_2 - \biggl[ v_1^2 - {2GM\over\ell_1} \biggr] = 0$.</h3>

#### Una vez que tenemos $v_2$ podemos calcular $\ell_2$  usando la relación $\ell_2 = \ell_1 v_1/v_2$.


#### **(b)**  Dados los valores de $v_1$, $\ell_1$, y $\ell_2$; otros parámetros de la órbita se obtienen mediante fórmulas simples que pueden derivarse de las leyes de Kepler y del hecho de que la órbita es una elipse:

<h4 align="center"><b>Semieje mayor</b>: $\; \; \; a = \frac{1}{2}(\ell_1+\ell_2),$</h3>

<h4 align="center"><b>Semieje menor</b>: $\; \; b = \sqrt{\ell_1\ell_2},$</h3>

<h4 align="center"><b>Período orbital</b>: $\; \; T = {2\pi ab\over\ell_1 v_1},$</h3>

<h4 align="center"><b>Excentricidad orbital</b>: $\; \; e = {\ell_2-\ell_1\over\ell_2+\ell_1},$</h3>

#### Escribe un programa que le pida al usuario que ingrese la distancia al Sol y la velocidad en el perihelio; para que calcule e imprima las cantidades $\ell_2$, $v_2$, $T$ y $e$.


In [6]:
import math

# Constantes
G = 6.6738e-11
M = 1.9891e30  # masa del Sol

# Entrada del usuario
l1 = float(input("Ingresa la distancia al Sol en el perihelio (en metros): "))
v1 = float(input("Ingresa la velocidad en el perihelio (en m/s): "))

# Coeficientes de la cuadrática
A = 1
B = - (2 * G * M) / (l1 * v1)
C = - (v1**2 - (2 * G * M) / l1)

# Resolver cuadrática
discriminante = B**2 - 4*A*C

if discriminante < 0:
    print("No hay solución real para v2.")
else:
    raiz = math.sqrt(discriminante)
    v2_1 = (-B + raiz) / (2*A)
    v2_2 = (-B - raiz) / (2*A)
    
    # Tomamos la raíz menor positiva
    if v2_1 < v2_2 and v2_1 > 0:
        v2 = v2_1
    else:
        v2 = v2_2

    # Calcular ℓ2
    l2 = l1 * v1 / v2

    # Semiejes
    a = 0.5 * (l1 + l2)
    b = math.sqrt(l1 * l2)

    # Periodo orbital
    T = (2 * math.pi * a * b) / (l1 * v1)

    # Excentricidad
    e = (l2 - l1) / (l2 + l1)

    # Imprimir resultados
    print("\nResultados orbitales:")
    print("ℓ2 =", l2, "m")
    print("v2 =", v2, "m/s")
    print("Periodo orbital T =", T, "s")
    print("Excentricidad e =", e)




Ingresa la distancia al Sol en el perihelio (en metros): 12
Ingresa la velocidad en el perihelio (en m/s): 76

Resultados orbitales:
ℓ2 = 14.25 m
v2 = 64.0 m/s
Periodo orbital T = 1.1824489427389866 s
Excentricidad e = 0.08571428571428572


#### **(c)** Prueba tu programa haciendo que calcule las propiedades de las órbitas de la Tierra (para las cuales $\ell_1=1.4710\times10^{11}\,$m y $v_1=3.0287\times10^4\,\mathrm{m\,s^{-1}}$)  y del cometa Halley ($\ell_1=8.7830\times10^{10}\,$m y $v_1=5.4529\times10^4\,\mathrm{m\,s^{-1}}$). Entre otras cosas, deberías encontrar que el período orbital de la Tierra es de un año y el del cometa Halley es de unos 76 años.

In [7]:
# Parte c): Tierra y Cometa Halley
print("\nValores para la Tierra:")
l1 = 1.4710e11
v1 = 3.0287e4

B = - (2 * G * M) / (l1 * v1)
C = - (v1**2 - (2 * G * M) / l1)
discriminante = B**2 - 4*A*C
v2 = (-B - math.sqrt(discriminante)) / (2*A)
l2 = l1 * v1 / v2
a = 0.5 * (l1 + l2)
b = math.sqrt(l1 * l2)
T = (2 * math.pi * a * b) / (l1 * v1)
e = (l2 - l1) / (l2 + l1)
print("T (Tierra) =", T / (3600*24*365), "años")

print("\nValores para el cometa Halley:")
l1 = 8.7830e10
v1 = 5.4529e4

B = - (2 * G * M) / (l1 * v1)
C = - (v1**2 - (2 * G * M) / l1)
discriminante = B**2 - 4*A*C
v2 = (-B - math.sqrt(discriminante)) / (2*A)
l2 = l1 * v1 / v2
a = 0.5 * (l1 + l2)
b = math.sqrt(l1 * l2)
T = (2 * math.pi * a * b) / (l1 * v1)
e = (l2 - l1) / (l2 + l1)
print("T (Halley) =", T / (3600*24*365), "años")



Valores para la Tierra:
T (Tierra) = 1.0002238777234564 años

Valores para el cometa Halley:
T (Halley) = 76.08170065465461 años


### 4. La fórmula semiempírica de la masa (FSM)

En física nuclear, la **fórmula de Weizsäcker** (conocida también como **fórmula semiempírica**) sirve para evaluar la masa y otras propiedades de un núcleo atómico, y está basada parcialmente en mediciones empíricas.
En particular, la fórmula se usa para calcular la **energía de enlace nuclear aproximada** \( B \), de un núcleo atómico con número atómico \( Z \) y número de masa \( A \):

$$
B = a_1 A - a_2 A^{2/3} - a_3 \frac{Z^2}{A^{1/3}} - a_4 \frac{(A - 2Z)^2}{A} + \frac{a_5}{\sqrt{A}}
$$

donde, en unidades de millones de electron-volts (MeV), las constantes son:

$$
a_1 = 15.8, \quad a_2 = 18.3, \quad a_3 = 0.714, \quad a_4 = 23.2
$$

y


$$
a_5 =
\begin{cases}
0 & \text{si } A \text{ es impar} \\
12.0 & \text{si } A \text{ y } Z \text{ son pares (ambos)} \\
-12.0 & \text{si } A \text{ es par y } Z \text{ impar}
\end{cases}
$$
---



a) Escribe un programa que tome como entrada los valores de A y Z, e imprima la 
energía de enlace B para el átomo correspondiente. Usa tu programa para encontrar
la energía de enlace de un átomo con A = 58 y Z = 28 (Hint: La respuesta
correcta es alrededor de los 490 MeV).


In [8]:
import math

# Constantes
a1 = 15.8
a2 = 18.3
a3 = 0.714
a4 = 23.2

# Entrada
A = int(input("Ingresa el número de masa A: "))
Z = int(input("Ingresa el número atómico Z: "))

# a5 depende de la paridad de A y Z
a5 = 0
if A % 2 != 0:
    a5 = 0
elif Z % 2 == 0:
    a5 = 12.0
else:
    a5 = -12.0

# Calcular B
B = a1*A - a2*A**(2/3) - a3*(Z**2)/A**(1/3) - a4*((A - 2*Z)**2)/A + a5/A**(1/2)
print("Energía de enlace B =", B, "MeV")



Ingresa el número de masa A: 67
Ingresa el número atómico Z: 43
Energía de enlace B = 306.6720254809317 MeV


b) Modifica el programa del inciso anterior, para escribir una segunda versión que
imprima no la energía de enlace total B, sino la energía de unión por nucleón, que
es B/A.


In [9]:
print("Energía de enlace por nucleón B/A =", B/A, "MeV")


Energía de enlace por nucleón B/A = 4.57719441016316 MeV


c) Escribe una tercera versión del programa para que tome como entrada solo un
valor del número atómico \( Z \) y luego pase por todos los valores de \( A \) desde \( A = Z \)
hasta \( A = 3Z \), para encontrar el que tiene la mayor energía de enlace por nucleón.
Este es el núcleo más estable con el número atómico dado. Haz que tu programa
imprima el valor de \( A \) para este núcleo más estable y el valor de la energía de
enlace por nucleón.

---


In [10]:
Z = int(input("Ingresa el número atómico Z para buscar A más estable: "))
A_mejor = Z
Bmax = 0

for A in range(Z, 3*Z + 1):
    if A % 2 != 0:
        a5 = 0
    elif Z % 2 == 0:
        a5 = 12.0
    else:
        a5 = -12.0
    B = a1*A - a2*A**(2/3) - a3*(Z**2)/A**(1/3) - a4*((A - 2*Z)**2)/A + a5/A**(1/2)
    BporA = B / A
    if BporA > Bmax:
        Bmax = BporA
        A_mejor = A

print("A más estable =", A_mejor, "con B/A =", Bmax, "MeV")


Ingresa el número atómico Z para buscar A más estable: 65
A más estable = 153 con B/A = 8.167786207290451 MeV


d) Finalmente, escribe una cuarta versión del programa que, en lugar de tomar \( Z \)
como entrada, se ejecute a través de todos los valores de \( Z \) de 1 a 100 e imprima
el valor más estable de \( A \) para cada uno. ¿A qué valor de \( Z \) se produce la energía
de enlace máxima por nucleón? (La respuesta correcta, en la vida real, es \( Z = 28 \),
que corresponde al Níquel).

### 5. La constante de Madelung

En física de la materia condensada, la **constante de Madelung** da el potencial eléctrico
total que siente un átomo en un sólido; y depende de las cargas de los otros átomos
cercanos y de sus ubicaciones.

Por ejemplo, el cristal de cloruro de sodio sólido (la sal de mesa), tiene átomos dispuestos
en una red cúbica, con átomos de sodio y cloro alternados, teniendo los de sodio
una carga positiva $+e$ y los de cloro una carga negativa $-e$ (donde $e$ es la carga del electrón).

Si etiquetamos cada posición en la red con tres coordenadas enteras $(i, j, k)$, entonces
los átomos de sodio caen en posiciones donde $i + j + k$ es par, y los átomos de cloro en
posiciones donde $i + j + k$ es impar.

Consideremos un átomo de sodio en el origen, i.e. $i = j = k = 0$, y calculemos la
constante de Madelung. Si el espaciado de los átomos en la red es $a$, entonces la distancia
desde el origen al átomo en la posición $(i, j, k)$ es:

$$
\sqrt{(ia)^2 + (ja)^2 + (ka)^2} = a \sqrt{i^2 + j^2 + k^2},
$$

y el potencial en el origen creado por tal átomo es:

$$
V(i, j, k) = \pm \frac{e}{4 \pi \varepsilon_0 a \sqrt{i^2 + j^2 + k^2}},
$$

siendo $\varepsilon_0$ la permitividad del vacío y el signo de la expresión se toma dependiendo de
si $i + j + k$ es par o impar.

Así entonces, el potencial total que siente el átomo de sodio
es la suma de esta cantidad sobre todos los demás átomos. Supongamos una caja cúbica
alrededor del átomo de sodio en el origen, con $L$ átomos en todas las direcciones, entonces:

$$
V_{\text{total}} = \sum_{\substack{i,j,k=-L \\ (i,j,k)\ne(0,0,0)}}^{L} V(i, j, k) = \frac{e}{4 \pi \varepsilon_0 a} M,
$$

donde $M$ es la constante de Madelung (al menos aproximadamente).

Técnicamente, la constante de Madelung es el valor de $M$ cuando $L \to \infty$, pero se
puede obtener una buena aproximación simplemente usando un valor grande de $L$.

**Escribe un programa para calcular e imprimir la constante de Madelung para el cloruro
de sodio. Utiliza un valor de $L$ tan grande como puedas, sin dejar que tu programa se
ejecute por más de un minuto.**

In [ ]:
import math

# Valor de L (cuanto mayor, más precisa la constante)
L = int(input("Ingresa el valor de L (por ejemplo, 10): "))

M = 0.0  # Aquí se acumulará la constante de Madelung

for i in range(-L, L+1):
    for j in range(-L, L+1):
        for k in range(-L, L+1):
            if i == 0 and j == 0 and k == 0:
                continue  # omitimos el átomo en el origen

            r = math.sqrt(i**2 + j**2 + k**2)

            if r != 0:
                signo = (-1)**(i + j + k)
                M += signo / r

print("\nAproximación de la constante de Madelung M =", M)


### 6. Coeficientes binomiales

El coeficiente binomial $\binom{n}{k}$ es un número entero igual a:

$$
\binom{n}{k} = \frac{n!}{k!(n - k)!} = \frac{n \times (n - 1) \times (n - 2) \times \dots \times (n - k + 1)}{1 \times 2 \times \dots \times k}
$$

donde $k \geq 1$, o bien $\binom{n}{0} = 1$ cuando $k = 0$.

**a)** Utiliza esta fórmula para escribir una función llamada `binomial(n, k)` (o como tú quieras) que calcule el coeficiente binomial para un $n$ y $k$ dados. Asegúrate de
que tu función devuelva la respuesta en forma de un número entero (no flotante)
y proporcione el valor correcto de $1$ para el caso en que $k = 0$.

In [13]:
# Entrada de n y k
n = int(input("Ingresa n: "))
k = int(input("Ingresa k: "))

# Calcular binomio sin usar funciones
num = 1
for i in range(n, n - k, -1):
    num *= i

den = 1
for i in range(1, k + 1):
    den *= i

binomio = num // den
print(f"El coeficiente binomial ({n} sobre {k}) es:", binomio)


Ingresa n: 9
Ingresa k: 3
El coeficiente binomial (9 sobre 3) es: 84


**b)** Usando tu función, escribe un programa que imprima las primeras 20 líneas del
“triángulo de Pascal”. La $n$-ésima línea del triángulo de Pascal contiene $n+1$ números,
que son los coeficientes $\binom{n}{0}$, $\binom{n}{1}$, y así sucesivamente hasta $\binom{n}{n}$. 

De tal manera que las primeras líneas son:

<pre>
               
             1   1
           1   2   1
         1   3   3   1
       1   4   6   4   1
    
</pre>


In [14]:
print("\nTriángulo de Pascal:")
for n in range(20):
    fila = ""
    for k in range(n + 1):
        num = 1
        for i in range(n, n - k, -1):
            num *= i
        den = 1
        for i in range(1, k + 1):
            den *= i
        fila += str(num // den) + " "
    print(fila)



Triángulo de Pascal:
1 
1 1 
1 2 1 
1 3 3 1 
1 4 6 4 1 
1 5 10 10 5 1 
1 6 15 20 15 6 1 
1 7 21 35 35 21 7 1 
1 8 28 56 70 56 28 8 1 
1 9 36 84 126 126 84 36 9 1 
1 10 45 120 210 252 210 120 45 10 1 
1 11 55 165 330 462 462 330 165 55 11 1 
1 12 66 220 495 792 924 792 495 220 66 12 1 
1 13 78 286 715 1287 1716 1716 1287 715 286 78 13 1 
1 14 91 364 1001 2002 3003 3432 3003 2002 1001 364 91 14 1 
1 15 105 455 1365 3003 5005 6435 6435 5005 3003 1365 455 105 15 1 
1 16 120 560 1820 4368 8008 11440 12870 11440 8008 4368 1820 560 120 16 1 
1 17 136 680 2380 6188 12376 19448 24310 24310 19448 12376 6188 2380 680 136 17 1 
1 18 153 816 3060 8568 18564 31824 43758 48620 43758 31824 18564 8568 3060 816 153 18 1 
1 19 171 969 3876 11628 27132 50388 75582 92378 92378 75582 50388 27132 11628 3876 969 171 19 1 


**c)** La probabilidad de que, para una moneda no sesgada lanzada $n$ veces, salga águila
$k$ veces es:

$$
p(k \mid n) = \frac{\binom{n}{k}}{2^n}
$$

Escribe un programa para calcular:

1) La probabilidad total de que una moneda lanzada 100 veces, salga águila exactamente 60 veces.

In [15]:
n = 100
k = 60

# Calcular binomio
num = 1
for i in range(n, n - k, -1):
    num *= i

den = 1
for i in range(1, k + 1):
    den *= i

binom = num // den
p = binom / 2**n
print("\nProbabilidad de obtener exactamente 60 águilas:", p)



Probabilidad de obtener exactamente 60 águilas: 0.010843866711637987


2) La probabilidad total de que salga águila **60 veces o más**.

In [16]:
prob = 0.0
for k in range(60, 101):
    num = 1
    for i in range(n, n - k, -1):
        num *= i
    den = 1
    for i in range(1, k + 1):
        den *= i
    binom = num // den
    prob += binom / 2**n

print("Probabilidad de obtener 60 o más águilas:", prob)


Probabilidad de obtener 60 o más águilas: 0.028443966820490392


### 7. Recursión

Una característica muy útil de las funciones que podemos definir es la **recursividad**, es
decir, la capacidad de una función de llamarse a sí misma. 

Por ejemplo, considere la siguiente definición del factorial $n!$ de un entero positivo $n$:

$$
n! =
\begin{cases}
1 & \text{si } n = 1, \\
n \times (n - 1)! & \text{si } n > 1.
\end{cases}
$$

Esto constituye una definición completa del factorial que nos permite calcular el valor
de $n!$ para cualquier entero positivo. Podemos emplear esta definición directamente
para crear una función para calcular factoriales, como esta (en Python):

```
def factorial(n):
    if n == 1:
        return 1
    else:
        return n * factorial(n - 1)
```
Nota cómo si $n$ no es igual a 1, la función se llama a sí misma para calcular el factorial
de $n - 1$, de tal manera que al pedir "print(factorial(5))" el intérprete de Python imprimirá correctamente la respuesta 120.

**a)** Anteriormente, en clase encontramos los números de Catalán $C_n$. Si hacemos una
pequeña modificación, la definición de $C_n$ se puede reescribir de la siguiente forma:

$$
C_n = 
\begin{cases}
1 & \text{if } n = 0, \\
\dfrac{4n - 2}{n + 1} C_{n-1} & \text{if } n > 0.
\end{cases}
$$

Escribe una función, usando recursividad, que calcule $C_n$ y utilízala para calcular
e imprimir $C_{100}$.

In [17]:
# Elegimos n = 100
n = 100
C = 1  # C₀ = 1

for i in range(1, n + 1):
    C = ((4 * i - 2) * C) // (i + 1)

print("C_100 =", C)


C_100 = 896519947090131496687170070074100632420837521538745909320


In [18]:
def catalan(n):
    if n == 0:
        return 1
    else:
        return ((4*n - 2) * catalan(n - 1)) // (n + 1)

print("C_100 =", catalan(100))  # Esto puede tardar mucho


C_100 = 896519947090131496687170070074100632420837521538745909320


**b)** Euclides demostró que el máximo común divisor $g(m, n)$ de dos enteros no negativos
$m$ y $n$ satisface que:

$$
g(m, n) =
\begin{cases}
m & \text{si } n = 0, \\
g(n, m \bmod n) & \text{si } n > 0.
\end{cases}
$$

Usando esta fórmula, escribe una función $g(m,n)$ que emplee recursividad para
calcular el máximo común divisor de $m$ y $n$. Usa tu función para calcular e imprimir
el máximo común divisor de 108 y 192.

In [19]:
m = 108
n = 192

while n != 0:
    r = m % n
    m = n
    n = r

print("Máximo común divisor (MCD) =", m)


Máximo común divisor (MCD) = 12


Comparando el cálculo de los números de Catalán del inciso 7a, con lo hecho en clase,
vemos que es posible hacer el cálculo de dos maneras, ya sea directamente o usando
recursividad.

En la mayoría de los casos, si una cantidad se puede calcular sin recursividad, será
más rápido y normalmente se recomienda hacerlo así, si es posible. 

Sin embargo, hay algunos cálculos que son esencialmente imposibles (o al menos mucho más difíciles)
sin recursividad.